### PR TITLE
fix(DataStore): drop failed constraint violation reconciliations

### PR DESCRIPTION
**Pull request by lawmicha**
*Issue #, if available:*

https://github.com/aws-amplify/amplify-ios/issues/1157 -> cascade delete fails for various reasons causing remote store to be missing the parent object. datastore full sync will fail in ReconcileAndLocalSaveOperation with a foreign key constraint violation when saving the child object. The ready event is not sent due to this failure.

https://github.com/aws-amplify/amplify-ios/issues/1310 -> also related to cascade delete failing, possible a bug in the cascade delete, however this fix will still be applicable to get the system in a ready state.

https://github.com/aws-amplify/amplify-ios/discussions/1097#discussioncomment-471914 -> additional context from @anatomybook. Sounds like "children without parents" scenario is bound to happen, not just in cascade delete. references #1001 

https://github.com/aws-amplify/amplify-ios/issues/1001 -> there are two bugs here. the first one is related to sync-ing models in the correct order but reconciling them concurrently which results in being done in the incorrect order. The first fix was done in #1128 with `ReconcileAndSaveQueue`. The first bug was fixed but the second bug still exists. This PR fix would address the second bug- now that models are reconcilled in the correct order, if there is a missing parent, there wil still be a constraint violation error

https://github.com/aws-amplify/amplify-android/pull/1058 -> Android PR for additional context

*Description of changes:*
`ReconcileAndLocalSaveOperation` will reconcile remote models from different sources (sync query results or subscription events). When the item is reconciled to be applied locally and the SQL operation fails with a constraint violation, the entire operation is fails. As a result of this, the `ready` event is never fired. By dropping the constraint violation errors and emitting a dropped model event, the `ready` event calculation will eventually be emitted once all models have been reconciled successfully (whether it was dropped or applied).

*Check points:*

- [ ] Added new tests to cover change, if needed
- [ ] All unit tests pass
- [ ] All integration tests pass
- [ ] Documentation update for the change if required
- [ ] PR title conforms to conventional commit style

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/Support/SQLiteResultError.swift [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import SQLite
+import Amplify
+
+/// Checks for specific SQLLite error codes
+/// See https://sqlite.org/rescode.html#primary_result_code_list for more details
+struct SQLiteResultError {
```
Comment (diegocstn) [inline AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/Support/SQLiteResultError.swift:13]: have you considered representing `SQLiteResultError` with an enum? And we can have a factory method to bridge a `DataStoreError` with it?

**Comment by diegocstn** [comment]
I think that we may need a level of abstraction/indirection between an `SQLiteError` and the `ReconcileAndLocalSaveOperation`. 
`ReconcileAndLocalSaveOperation` at that point doesn't have any knowledge about the implementation of the `StorageAdapter` (which in this case happens to be a SQLite database). 

Can we perhaps have an associated error type on `StorageEngineAdapter` or a `shouldIgnoreError` method (or similar) that storage classes conforming to `StorageEngineAdapter` need to implement?

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/Support/SQLiteResultError.swift (hunk 1 above)
Comment (lawmicha) [reply AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/Support/SQLiteResultError.swift:13]: example
```
extension DataStoreError
  static func toSQLiteError() -> SQLiteError? 
}

enum SQLiteError {
  case constraintViolation
  case other(message, code)

  init(error: DataStoreError) throw {
  }
}
```

**Comment by anatomybook** [comment]
Hello,

could you kindly provide some insight as to when this will go onto main?

Thanks so much!

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Storage/ModelStorageBehavior.swift [hunk 2]
```diff
@@ -44,3 +44,13 @@ protocol ModelStorageBehavior {
                          completion: DataStoreCallback<[M]>)
 
 }
+
+protocol ModelStorageErrorBehavior {
+    func shouldIgnoreError(error: DataStoreError) -> Bool
```
Comment (diegocstn) [inline AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Storage/ModelStorageBehavior.swift:49]: do you think conforming classes may also need the operation type as additional parameter here? 
I wonder that we may need to treat an error differently based on the context where it originated

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/Support/SQLiteResultError.swift [hunk 3]
```diff
@@ -9,20 +9,33 @@ import SQLite
...
+    case constraintViolation(statement: Statement?)
+
+    /// Represents a SQLite specific [error code](https://sqlite.org/rescode.html)
+    ///
+    /// - message: English-language text that describes the error
+    /// - code: SQLite [error code](https://sqlite.org/rescode.html#primary_result_code_list)
+    /// - statement: the statement which produced the error
+    case error(message: String, code: Int32, statement: Statement?)
+
+    init?(dataStoreError: DataStoreError) {
```
Comment (diegocstn) [inline AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/Support/SQLiteResultError.swift:27]: nit: maybe `init?(from dataStoreError: DataStoreError) {`?

**Review by diegocstn — APPROVED** [review]

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Storage/ModelStorageBehavior.swift (hunk 2 above)
Comment (lawmicha) [reply AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Storage/ModelStorageBehavior.swift:49]: what's the operation type you are referring to here?

Comment (diegocstn) [reply AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Storage/ModelStorageBehavior.swift:49]: I was referring to the "operation types" defined by `ModelStorageBehavior` like `query`, `save`, `delete` (but I don't think we have an actual Swift type for those)

Comment (lawmicha) [reply AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Storage/ModelStorageBehavior.swift:49]: ah got it, im not sure if the decision to ignore an error differs based on operation type. we can keep this in mind if the use case extends beyond constraint violation errors

**Review by diegocstn — APPROVED** [review]

**Pull request merged**